### PR TITLE
Kernel#sleep and Mutex#sleep can simply call Scheduler#block

### PR DESCRIPTION
**Pull request by eregon**
* Remove the no-longer-needed `#kernel_sleep` callback.

### CODE REVIEW

File: thread_sync.c [hunk 1]
```diff
@@ -535,8 +541,7 @@ rb_mutex_sleep(VALUE self, VALUE timeout)
 
     VALUE scheduler = rb_scheduler_current();
     if (scheduler != Qnil) {
-        rb_scheduler_kernel_sleep(scheduler, timeout);
-        mutex_lock_uninterruptible(self);
+        rb_ensure(call_rb_scheduler_block_mutex_sleep, timeout, mutex_lock_uninterruptible, self);
```
Comment (eregon) [inline thread_sync.c:544]: This ensure seems important and follows the pattern below.
Whatever `call_rb_scheduler_block_mutex_sleep()` does, we need to reacquire the Mutex when getting out of `Mutex#sleep`.

**Comment by ioquatix** [comment]
I need to think about this.

**Comment by eregon** [comment]
@ioquatix What's the motivation to keep `kernel_sleep`?
One can still check if it's `Kernel#sleep`, `Mutex#sleep` or some other blocking call by checking the first argument.
Also currently `Scheduler#kernel_sleep` is called for `Mutex#sleep` so that seems confusing.

**Comment by ioquatix** [comment]
I'm not convinced by this change.

We expect normal flow to be `block` followed by `unblock` and timeout should be exceptional case.

For `kernel_sleep`, waiting for the timeout is normal case, and returning early is exceptional case.

In many implementations, `kernel_sleep` would be the similar implementation as `block` but that is not strong enough for me to want to change the interface to directly call `block`, because as we see, `block(:sleep, timeout)` is sending a symbol as `blocker` and that's also not quite right.

**Comment by eregon** [comment]
> I'm not convinced by this change.

It's an obvious simplification. Simplifying the interface is good, isn't it?

> We expect normal flow to be `block` followed by `unblock` and timeout should be exceptional case.
> 
> For `kernel_sleep`, waiting for the timeout is normal case, and returning early is exceptional case.

I would say using `sleep` is rare in real programs.
In any case, I don't think the frequency of operations matter, they must be supported by the scheduler anyway.

I don't think these expectations lead to anything useful in practice, so IMHO they don't matter.

> `block(:sleep, timeout)` is sending a symbol as `blocker` and that's also not quite right either.

What's an actual issue about that?

**Comment by eregon** [comment]
@ioquatix Could you make your mind about this?

**Comment by ioquatix** [comment]
Okay, I'm against this change.

File: thread_sync.c (hunk 1 above)
Comment (ioquatix) [reply thread_sync.c:544]: I think we should make a separate PR to address this if still necessary.

Comment (eregon) [reply thread_sync.c:544]: I'll let you integrate this fix since you rejected this PR.

**Comment by eregon** [comment]
Meh, now we have a method called `kernel_sleep` which is called by `Mutex#sleep`, that seems confusing.
Anyway, it's your call, better reject it than leave it open forever.

File: thread_sync.c (hunk 1 above)
Comment (ioquatix) [reply thread_sync.c:544]: Ok.

**Pull request closed without merging**